### PR TITLE
Allow for cropping transparent true color pngs

### PR DESCRIPTION
**Pull request by Reettaphant**
Allow cropping pngs with transparency. Optimised the generated assets with pngquant. 

Getting this working working locally requires installing pngquant. If installing pngquant with brew, create a symlink from /usr/local/bin/pngquant to /urs/bin/pngquant

### CODE REVIEW

File: common-lib/src/main/scala/com/gu/mediaservice/lib/aws/S3.scala [hunk 1]
```diff
@@ -37,9 +37,15 @@ class S3(credentials: AWSCredentials) {
   }
 
   private def getContentDispositionFilename(image: Image, charset: CharSet): String = {
+
+    val extension = image.source.mimeType match {
+      case Some("image/jpeg") => "jpg"
```
Comment (kenoir) [inline common-lib/src/main/scala/com/gu/mediaservice/lib/aws/S3.scala]: Might be better to identify png specifically here and fail in the "other" case.

File: common-lib/src/main/scala/com/gu/mediaservice/lib/imaging/ImageOperations.scala [hunk 2]
```diff
@@ -78,18 +79,19 @@ object ImageOperations {
   }
 
   def cropImage(sourceFile: File, bounds: Bounds, qual: Double = 100d, tempDir: File,
-                iccColourSpace: Option[String], colourModel: Option[String]): Future[File] = {
+                iccColourSpace: Option[String], colourModel: Option[String], fileType: String): Future[File] = {
```
Comment (kenoir) [inline common-lib/src/main/scala/com/gu/mediaservice/lib/imaging/ImageOperations.scala]: Do you think it's worth using an enum or set of case classes (with sealed trait) for MimeType?

File: kahuna/public/js/edits/image-editor.html [hunk 3]
```diff
@@ -70,7 +70,7 @@
                     <gr-icon title="You need to add both a credit and description, (PNGs are also invalid).">help</gr-icon>
                 </span>
                 <div class="result-editor__info-item--png-warning"ng:if="ctrl.invalidReasons['is_invalid_png']">
-                    <strong>PNG</strong>: Pngs with transparency not supported!
+                    <strong>PNG</strong>: This type currently not supported.
```
Comment (kenoir) [inline kahuna/public/js/edits/image-editor.html]: Maybe we could add a title tip here with some more information about what's wrong, how to fix it?

**Comment by kenoir** [comment]
:+1: apart from comments.

Let me know if you want to talk through the enum/case class stuff.

File: common-lib/src/main/scala/com/gu/mediaservice/lib/imaging/ImageOperations.scala (hunk 2 above)
Comment (Reettaphant) [inline common-lib/src/main/scala/com/gu/mediaservice/lib/imaging/ImageOperations.scala]: Yes, I think that's a good idea. I've added a sealed trait for cropping images now.

**Comment by Reettaphant** [comment]
@NickPapacostas This has now been updated and is ready for review

File: cropper/app/lib/Crops.scala [hunk 4]
```diff
@@ -26,14 +27,27 @@ object Crops {
...
     val metadata = apiImage.metadata
     val iccColourSpace = FileMetadataHelper.normalisedIccColourSpace(apiImage.fileMetadata)
 
     for {
       strip <- ImageOperations.cropImage(sourceFile, source.bounds, 100d, Config.tempDir, iccColourSpace, colourModel, mediaType.extension)
-      file  <- ImageOperations.appendMetadata(strip, metadata)
+      file: File <- ImageOperations.appendMetadata(strip, metadata)
+
+      //Before apps and frontend can handle PNG24s we need to pngquant PNG24 master crops
+      optimisedFile =  if (colourType == "True Color" || colourType == "True Color with Alpha") {
```
Comment (paperboyo) [inline cropper/app/lib/Crops.scala]: Could it ever be `True Color`? I thought these would go through PNGs-that-should-really-be-JPGs route... Not a problem as such, though.

File: cropper/app/lib/Crops.scala [hunk 5]
```diff
@@ -21,35 +23,53 @@ object Crops {
...
       for {
-        file       <- ImageOperations.resizeImage(sourceFile, dimensions, 75d, Config.tempDir)
-        filename    = outputFilename(apiImage, crop.specification.bounds, dimensions.width)
-        sizing     <- CropStore.storeCropSizing(file, filename, mediaType, crop, dimensions)
-        _          <- delete(file)
+        file          <- ImageOperations.resizeImage(sourceFile, dimensions, 75d, Config.tempDir, mediaType.extension)
+        optimisedFile = ImageOperations.optimiseImage(file, mediaType)
+        filename      = outputFilename(apiImage, crop.specification.bounds, dimensions.width, mediaType.extension)
+        sizing       <- CropStore.storeCropSizing(optimisedFile, filename, mediaType.extension, crop, dimensions)
+        _            <- delete(file)
```
Comment (NickPapacostas) [inline cropper/app/lib/Crops.scala]: dont we just need to delete optimisedFile? (cus it just equals file if it's not one of the special png types)

File: cropper/app/lib/Crops.scala (hunk 4 above)
Comment (Reettaphant) [inline cropper/app/lib/Crops.scala]: Yes, I think I forgot all about that. I'll double check and fix it.

File: kahuna/public/js/services/image/downloads.js [hunk 6]
```diff
@@ -16,12 +16,13 @@ imageDownloadsService.factory('imageDownloadsService', ['imgops', '$http', funct
...
         const imageId = imageData.id;
+        const extension = imageData.source.mimeType === 'image/jpeg' ? 'jpg' : 'png';
 
         if (filename) {
             const basename = stripExtension(filename);
-            return `${basename} (${imageId}).jpg`;
+            return `${basename} (${imageId}).` + extension;
         } else {
-            return `${imageId}.jpg`;
+            return `${imageId}.jpg` + extension;
```
Comment (NickPapacostas) [inline kahuna/public/js/services/image/downloads.js]: gotta remove ".jpg" i think

File: cropper/app/lib/Crops.scala (hunk 5 above)
Comment (Reettaphant) [inline cropper/app/lib/Crops.scala]: I think we probably have to delete both if they exist.

**Comment by NickPapacostas** [comment]
Nice job! 

A few little cleanups, then 👍

File: cropper/app/lib/Crops.scala (hunk 5 above)
Comment (NickPapacostas) [inline cropper/app/lib/Crops.scala]: 👍

**Pull request merged**